pretrain/openwebtext.py

```diff
@@ -97,8 +98,10 @@ def main(fabric, resume) -> None:
     )
     model, optimizer = fabric.setup(model, optimizer)
 
-    fabric.seed_everything(1337 + fabric.global_rank)
-    train_data, val_data = load_datasets(data_dir)
+    train_data, val_data = load_datasets(data_dir, block_size=model.config.block_size)
+    train_dataloader = DataLoader(train_data, batch_size=micro_batch_size, num_workers=2)
+    val_dataloader = DataLoader(val_data, batch_size=micro_batch_size, num_workers=2)
```

Comment:
Just out of curiosity: why limit to 2 workers? More workers don't bring any significant improvement?
Maybe it's better specify the number of workers automatically? Maybe take 80% of available cores (with 100% it might be actually slower if the system wants to do something else and all cores are occupied).

```python
NUM_WORKERS = int(multiprocessing.cpu_count() * 0.8)
```
or if to count only physical cores:
```python
NUM_WORKERS = int(psutil.cpu_count(logical=False) * 0.8)
```

Reply:
Hi @Andrei-Aksionov 
I determined this experimentally, that's all. 2 workers per GPU seemed enough for my experiments and monitored with the pytorch profiler. 

I think an automatic formula could be done, but I suggest to incorporate the local world size of processes (num devices set in Fabric).

Reply:
Unfortunately I don't have experience with training on multiple GPUs. Please correct me if I'm wrong.
As I understand for a server with a single CPU and multiple GPUs an instance of `DataLoader` will be created for each GPU so we need to split available cores (or 80-90% of them) of the single CPU across all GPUs. And make sure that there are at least 2 workers.

```python
NUM_WORKERS = max(int(psutil.cpu_count(logical=False) * 0.9 / fabric.world_size), 2)
```

Reply:
Yes, I would say that this is a viable approach for the general case. Usually, more workers are needed for training vision models where extensive data augmentations can benefit from multiprocessing. In our case, the data loading is only accessing the disk and not doing any heavy computations otherwise, which means that doing it in the main process is just slow enough to make the GPU wait and dips utilization, but adding more than 1-2 workers doesn't benefit us more because loading them in parallel to the GPU work is already solving the issue.

Feel free to open an issue or a PR with your proposal, it doesn't hurt. A small correction in the formula would be:
`NUM_WORKERS = max(int(psutil.cpu_count(logical=False) * 0.9 / devices), 2)` because in multi-node training, each machine has its own CPU so the count should be divided by the number of local processes.

Reply:
Thanks @awaelchli for the explanation.
Now it looks for me as an unnecessary complication.
So I'll not create a PR for it. KISS and all that jazz.

Reply:
I appreciate the review @Andrei-Aksionov, good points <3